server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierTimeTracker.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.service;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.metrics.CounterMetric;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.core.TimeValue;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public final class ClusterApplierTimeTracker {
+
+    // Most listeners are added when the node is initialized, however a few
+    // are added during the lifetime of a node and for these cases we need ConcurrentHashMap
+    private final Map<String, CounterMetric> timeSpentPerApplier = new ConcurrentHashMap<>();
+    private final Map<String, CounterMetric> timeSpentPerListener = new ConcurrentHashMap<>();
+
+    void addApplier(String name) {
+        timeSpentPerApplier.put(name, new CounterMetric());
+    }
+
+    void removeApplier(String name) {
+        timeSpentPerApplier.remove(name);
+    }
+
+    void addListener(String name) {
+        timeSpentPerListener.put(name, new CounterMetric());
+    }
+
+    void removeListener(String name) {
+        timeSpentPerListener.remove(name);
```

Comment:
It'd be good to keep track of the accumulated time spent on removed listeners somewhere, just so we can see a bit better what's happening if there's a listener/applier which gets added and removed a lot which is slow.

Reply:
So you mean an entry in the stats that represents time spent for all listeners that were removed?